File: backend/app/routers/access.py
from fastapi import APIRouter, Depends, Header
from sqlalchemy.orm import Session

from app.access_keys import gate_active, is_admin_key, is_valid_key
from app.db import get_db
from app.schemas import AccessStatusRead

router = APIRouter(prefix="/access", tags=["access"])


@router.get("/status", response_model=AccessStatusRead)
def access_status(
    x_unlock_key: str | None = Header(default=None),
    db: Session = Depends(get_db),
) -> AccessStatusRead:
    """Let the UI show a lock instead of firing a doomed request: `locked` is
    whether the import gate is enforcing, `unlocked` whether this client's key
    gets through it (ADR-009).

    `admin` is a third, independent answer (#354): whether this key may read
    everyone's diagnostics. It is **not** implied by `unlocked` — every tester
    is unlocked and almost none are administrators — which is the distinction
    that makes the diagnostics page gate on this field rather than on that one.
    """
    locked = gate_active(db)
    unlocked = not locked or (x_unlock_key is not None and is_valid_key(db, x_unlock_key))
    return AccessStatusRead(locked=locked, unlocked=unlocked, admin=is_admin_key(db, x_unlock_key))

File: frontend/src/api/access.ts
import { useQuery } from '@tanstack/react-query'

import { apiFetch } from './client'

export interface AccessStatus {
  /** Whether importing is gated on this server at all. */
  locked: boolean
  /** Whether this client's stored key currently gets through the gate. */
  unlocked: boolean
  /**
   * Whether this key may read everyone's diagnostics (#354).
   *
   * **Independent of `unlocked`, not implied by it.** Every invited tester is
   * unlocked and almost none are administrators — which is exactly why the
   * diagnostics page gates on this and not on that.
   */
  admin: boolean
}

export const ACCESS_STATUS_KEY = ['access-status'] as const

export function useAccessStatus() {
  return useQuery({
    queryKey: ACCESS_STATUS_KEY,
    queryFn: () => apiFetch<AccessStatus>('/access/status'),
  })
}
